### PR TITLE
Adds branch coverage measurement

### PR DESCRIPTION
**Pull request by jaltmayerpizzorno**

### CODE REVIEW

File: pyptr.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        Py_IncRef((PyObject*)o);
+        return PyPtr(o);
+    }
+
+    O* operator->() { return _obj; }
+
+    operator O*() { return _obj; }
+
+
+    PyPtr& operator=(const PyPtr& ptr) {
```
Comment (emeryberger) [inline pyptr.h:30]: This is stale - the Scalene version does an equality test, and I think has more comments as well.

File: slipcover/__main__.py [hunk 2]
```diff
@@ -3,16 +3,17 @@
 from typing import Any, Dict
 from slipcover import slipcover as sc
 from slipcover import bytecode as bc
+from slipcover import branch as br
```
Comment (emeryberger) [inline slipcover/__main__.py:6]: consider adding `isort` to your builds to sort the imports.

File: slipcover/__main__.py [hunk 3]
```diff
@@ -3,16 +3,17 @@
...
+import ast
 import atexit
 
-
 from importlib.abc import MetaPathFinder, Loader
-from importlib.util import spec_from_loader
 
 class SlipcoverLoader(Loader):
-    def __init__(self, sci, orig_loader):
+    def __init__(self, sci, orig_loader, origin):
```
Comment (emeryberger) [inline slipcover/__main__.py:13]: Please add comments for each field in `self` as well as type annotations.

File: slipcover/__main__.py [hunk 4]
```diff
@@ -29,11 +30,19 @@ def get_code(self, name):   # expected by pyrun
         return self.orig_loader.get_code(name)
 
     def exec_module(self, module):
-        code = self.orig_loader.get_code(module.__name__)
+        import importlib.machinery
+        if sci.branch and isinstance(self.orig_loader, importlib.machinery.SourceFileLoader) and self.origin.exists():
```
Comment (emeryberger) [inline slipcover/__main__.py:34]: Could use a higher level comment describing what this logic is doing.

File: slipcover/__main__.py [hunk 5]
```diff
@@ -65,6 +74,7 @@ def find_spec(self, fullname, path, target=None):
 #
 import argparse
 ap = argparse.ArgumentParser(prog='slipcover')
+ap.add_argument('--branch', action='store_true', help="additionally measure branch coverage")
```
Comment (emeryberger) [inline slipcover/__main__.py:77]: change help to `measure both branch and line coverage`

File: slipcover/__main__.py [hunk 6]
```diff
@@ -118,13 +128,37 @@ def exec_wrapper(obj, g):
...
+        return None
+
     for f in sc.Slipcover.find_functions(_pytest.assertion.rewrite.__dict__.values(), set()):
         if 'exec' in f.__code__.co_names:
             ed = bc.Editor(f.__code__)
             wrapper_index = ed.add_const(exec_wrapper)
             ed.replace_global_with_const('exec', wrapper_index)
             f.__code__ = ed.finish()
 
+        if sci.branch and 'rewrite_asserts' in f.__code__.co_names:
```
Comment (emeryberger) [inline slipcover/__main__.py:149]: some more comments here, please

File: slipcover/bytecode.py [hunk 7]
```diff
@@ -43,11 +43,13 @@ def branch2offset(arg: int) -> int:
...
 op_JUMP_FORWARD = dis.opmap["JUMP_FORWARD"]
 op_NOP = dis.opmap["NOP"]
+op_STORE_NAME = dis.opmap["STORE_NAME"]
+op_STORE_GLOBAL = dis.opmap["STORE_GLOBAL"]
 
 
 def arg_ext_needed(arg: int) -> int:
     """Returns the number of EXTENDED_ARGs needed for an argument."""
-    return (arg.bit_length() - 1) // 8
+    return -((arg>>8).bit_length() // -8)   # ceiling by way of -(a // -b)
```
Comment (emeryberger) [inline slipcover/bytecode.py:52]: Pretty opaque stuff here - is it performance sensitive?

File: slipcover/bytecode.py [hunk 8]
```diff
@@ -466,9 +469,15 @@ def add_const(self, value):
...
 
 
-    def insert_function_call(self, offset, function, args):
-        """Inserts a function call."""
+    def insert_function_call(self, offset, function, args, repl_length=0):
+        """Inserts a function call.
 
+        *repl_length*, if passed, indicates the number of bytes to replace at
+        that offset.  The inserted bytecode must currently be at least as
+        large as *repl_length*.
```
Comment (emeryberger) [inline slipcover/bytecode.py:477]: So really, shouldn't it be `repl_min_length`?

File: slipcover/bytecode.py [hunk 9]
```diff
@@ -504,22 +513,45 @@ def insert_function_call(self, offset, function, args):
...
         len_insert = len(insert)
+        assert len_insert >= repl_length
 
         insert[1] = offset2branch(len_insert-2)    # fails if > 255
         self.max_addtl_stack = max(self.max_addtl_stack, calc_max_stack(insert))
 
-        self.patch[offset:offset] = insert
+        self.patch[offset:offset+repl_length] = insert
+
+        change = len_insert - repl_length
```
Comment (emeryberger) [inline slipcover/bytecode.py:523]: Use a more descriptive name than `change`.

File: slipcover/bytecode.py [hunk 10]
```diff
@@ -572,7 +607,7 @@ def replace_inserted_function(self, offset, new_func_index):
 
         assert op == op_LOAD_CONST
 
-        # FIXME use actual argument lenth rather than art_ext_needed
+        # FIXME use actual argument length rather than arg_ext_needed
```
Comment (emeryberger) [inline slipcover/bytecode.py:610]: Can you explain why this needs to be FIXed?

File: slipcover/bytecode.py [hunk 11]
```diff
@@ -629,44 +665,56 @@ def find_load_globals():
...
+    def _finish(self):
+        if not self.finished:
+            self.finished = True
 
-        assert not self.finished
-        self.finished = True
+            if self.branches is not None:
+                # A branch's new target may now require more EXTENDED_ARG opcodes to be expressed.
+                # Inserting space for those may in turn trigger needing more space for others...
+                # FIXME missing test for length adjustment triggering other length adjustments
```
Comment (emeryberger) [inline slipcover/bytecode.py:675]: Correctness issue?

File: slipcover/slipcover.py [hunk 12]
```diff
@@ -81,9 +82,10 @@ def matches(self, filename : Path):
 
 
 class Slipcover:
-    def __init__(self, collect_stats : bool = False, d_threshold = 50):
+    def __init__(self, collect_stats: bool = False, d_threshold: int = 50, branch: bool = False):
```
Comment (emeryberger) [inline slipcover/slipcover.py:85]: What's the`d_threshold` for? Could use a more descriptive name, no doubt.

**Review by emeryberger — APPROVED** [review]

File: pyptr.h (hunk 1 above)
Comment (jaltmayerpizzorno) [reply pyptr.h:30]: I merged the scalene changes.

File: slipcover/__main__.py (hunk 2 above)
Comment (jaltmayerpizzorno) [reply slipcover/__main__.py:6]: Mm, thanks.

File: slipcover/__main__.py (hunk 3 above)
Comment (jaltmayerpizzorno) [reply slipcover/__main__.py:13]: Type annotations aren't present in `importlib.abc.Loader`...  I've guessed what I could, but left many thinking none are better than incorrect ones.

File: slipcover/bytecode.py (hunk 8 above)
Comment (jaltmayerpizzorno) [reply slipcover/bytecode.py:477]: No... the intention is that you pass in `repl_length` asking it to replace that many bytes. The 2nd sentence was just documenting a current limitation.

File: slipcover/bytecode.py (hunk 10 above)
Comment (jaltmayerpizzorno) [reply slipcover/bytecode.py:610]: We want our replacement code to have the same length, so to use the same number of `EXTENDED_ARG` opcodes; but `arg_ext_needed` indicate the minimum needed to encode an argument, not how many there actually were (and we're replacing). I'll improve the comment.

Comment (jaltmayerpizzorno) [reply slipcover/bytecode.py:610]: Usually they are the same, but it is _possible_, if not likely, that unnecessary `EXTENDED_ARG`s were used.

File: slipcover/bytecode.py (hunk 11 above)
Comment (jaltmayerpizzorno) [reply slipcover/bytecode.py:675]: I am not sure what you are asking... this code replicates what I've seen Python code do, but I didn't put enough time into creating tests to successfully test out this scenario. That is what the FIXME is saying... that this isn't tested out.

File: slipcover/slipcover.py (hunk 12 above)
Comment (jaltmayerpizzorno) [reply slipcover/slipcover.py:85]: Thanks. `d_miss_threshold` to trigger de-instrumentation.

File: slipcover/bytecode.py (hunk 7 above)
Comment (jaltmayerpizzorno) [reply slipcover/bytecode.py:52]: It was the incorrect calculation for some cases.  I added tests and the bug became apparent.
I tried reverting, it's not what caused the 2% slowdown.

**Pull request merged**